### PR TITLE
fix: add a post/response/comment during active blackout dates fixed for all roles

### PR DESCRIPTION
**Pull request by ayeshoali**
[INF-628](https://2u-internal.atlassian.net/browse/INF-628)
Add a post/response/comment during active blackout dates fixed for all roles in discussions MFE
**Before**
For discussions MFE when blackout dates were active, buttons for adding a post, response, or comment were not visible for some roles.
**After**
For discussions MFE when blackout dates are active, buttons for adding a post, response, or comment are visible for roles as seen below.

- Global staff
- Course staff 
- Course admin
- Disc. admin
- Disc. moderator
- Disc. TA 
- Disc. Group TA

### CODE REVIEW

File: src/discussions/comments/comment/Comment.jsx [hunk 1]
```diff
@@ -39,6 +42,24 @@ function Comment({
...
   const blackoutDateRange = useSelector(selectBlackoutDate);
+  const isUserAdmin = useSelector(selectUserIsStaff);
+  const userHasModerationPrivilages = useSelector(selectUserHasModerationPrivileges);
+  const isUserGroupTA = useSelector(selectUserIsGroupTa);
+  const isCourseAdmin = useSelector(selectIsCourseAdmin);
+  const isCourseStaff = useSelector(selectIsCourseStaff);
+
+  const hasPrivilege = handleAddPostForRoles(isUserAdmin, userHasModerationPrivilages,
+    isUserGroupTA, isCourseAdmin, isCourseStaff);
+
```
Comment (mehaknasir) [inline src/discussions/comments/comment/Comment.jsx:53]: this utility function can be changed to custom hook and all selectors then can easily be utilized there and hook will be used at all placed needed to calculate user privilege.

File: src/discussions/comments/comment/ResponseEditor.jsx [hunk 2]
```diff
@@ -20,12 +23,29 @@ function ResponseEditor({
...
+  const isCourseStaff = useSelector(selectIsCourseStaff);
 
   useEffect(() => {
     setAddingResponse(false);
   }, [postId]);
 
-  const blackoutDateRange = useSelector(selectBlackoutDate);
+  const hasPrivilege = handleAddPostForRoles(isUserAdmin, userHasModerationPrivilages,
+    isUserGroupTA, isCourseAdmin, isCourseStaff);
+
```
Comment (mehaknasir) [inline src/discussions/comments/comment/ResponseEditor.jsx:39]: this duplication will be removed when you shift the logic in custom hook

File: src/discussions/posts/post-actions-bar/PostActionsBar.jsx [hunk 3]
```diff
@@ -26,11 +31,29 @@ function PostActionsBar({
...
+  const isCourseAdmin = useSelector(selectIsCourseAdmin);
+  const isCourseStaff = useSelector(selectIsCourseStaff);
 
   const handleCloseInContext = () => {
     postMessageToParent('learning.events.sidebar.close');
   };
 
+  const hasPrivilege = handleAddPostForRoles(isUserAdmin, userHasModerationPrivilages,
+    isUserGroupTA, isCourseAdmin, isCourseStaff);
+
```
Comment (mehaknasir) [inline src/discussions/posts/post-actions-bar/PostActionsBar.jsx:46]: same here

**Review by mehaknasir — CHANGES_REQUESTED** [review]
code architectural change is required also test case need to be written based on change in use case.

File: src/discussions/utils.js [hunk 4]
```diff
@@ -281,3 +281,11 @@ export function inBlackoutDateRange(blackoutDateRanges) {
...
   );
 }
+
+export function handleAddPostForRoles(isUserAdmin, userHasModerationPrivilages, isUserGroupTA,
+  isCourseAdmin, isCourseStaff) {
+  if (isUserAdmin || userHasModerationPrivilages || isUserGroupTA || isCourseAdmin || isCourseStaff) {
+    return true;
+  }
+  return false;
+}
```
Comment (awais-ansari) [inline src/discussions/utils.js:291]: Move this code into useUserPrivilage hook and refactor it.
return ((isUserAdmin || userHasModerationPrivilages || isUserGroupTA || isCourseAdmin || isCourseStaff) || false )

File: src/discussions/data/hooks.js [hunk 5]
```diff
@@ -174,3 +178,16 @@ export const useCurrentDiscussionTopic = () => {
...
+  const userHasModerationPrivilages = useSelector(selectUserHasModerationPrivileges);
+  const isUserGroupTA = useSelector(selectUserIsGroupTa);
+  const isCourseAdmin = useSelector(selectIsCourseAdmin);
+  const isCourseStaff = useSelector(selectIsCourseStaff);
+
+  return (
+    handleAddPostForRoles(isUserAdmin, userHasModerationPrivilages,
+      isUserGroupTA, isCourseAdmin, isCourseStaff)
+  );
+};
```
Comment (awais-ansari) [inline src/discussions/data/hooks.js:193]: rename this hook. The current name is confusing. 
like useUserCanAddThreadInBlackoutDate

**Review by awais-ansari — CHANGES_REQUESTED** [review]

File: src/discussions/posts/post-actions-bar/PostActionsBar.jsx [hunk 6]
```diff
@@ -26,11 +27,22 @@ function PostActionsBar({
...
+  const handleAddPost = () => {
+    if (loadingStatus === RequestStatus.SUCCESSFUL && (!(inBlackoutDateRange(blackoutDateRange)))) {
+      return true;
+    }
+    if (hasPrivilege) {
+      return true;
+    }
+    return false;
+  };
+
```
Comment (awais-ansari) [inline src/discussions/posts/post-actions-bar/PostActionsBar.jsx:45]: You will not need redundant this check after the hook modification.  The hook will simply return a Boolean value.

**Review by awais-ansari — CHANGES_REQUESTED** [review]

File: src/discussions/data/hooks.js [hunk 7]
```diff
@@ -174,3 +179,18 @@ export const useCurrentDiscussionTopic = () => {
...
+  const isUserGroupTA = useSelector(selectUserIsGroupTa);
+  const isCourseAdmin = useSelector(selectIsCourseAdmin);
+  const isCourseStaff = useSelector(selectIsCourseStaff);
+  if (!inBlackoutDateRange(blackoutDateRange)) {
+    return true;
+  }
+  return (
+    isUserAdmin || userHasModerationPrivilages || isUserGroupTA || isCourseAdmin || isCourseStaff
+  );
+};
```
Comment (awais-ansari) [inline src/discussions/data/hooks.js:196]: refactor this condition.

**Review by awais-ansari — CHANGES_REQUESTED** [review]
Add test case for this use case.

**Review by mehaknasir — COMMENTED** [review]
Test cases need to be written for the use cases, would be good if you write test cases for the hook and it will be covering all the parts where the hook is used.

File: src/discussions/data/hooks.js [hunk 8]
```diff
@@ -174,3 +179,17 @@ export const useCurrentDiscussionTopic = () => {
...
 };
+
+export const useUserCanAddThreadInBlackoutDate = () => {
+  const blackoutDateRange = useSelector(selectBlackoutDate);
+  const isUserAdmin = useSelector(selectUserIsStaff);
+  const userHasModerationPrivilages = useSelector(selectUserHasModerationPrivileges);
+  const isUserGroupTA = useSelector(selectUserIsGroupTa);
+  const isCourseAdmin = useSelector(selectIsCourseAdmin);
+  const isCourseStaff = useSelector(selectIsCourseStaff);
+  const ifInBlackoutDateRange = inBlackoutDateRange(blackoutDateRange);
```
Comment (awais-ansari) [inline src/discussions/data/hooks.js:190]: Rema the const `ifInBlackoutDateRange` to `IsInBlackoutDateRange`

Comment (awais-ansari) [reply src/discussions/data/hooks.js:190]: date T, Role T => Add T
date F, Role T => Add T
date T, Role F => Add F
date F, Role F => Add T

!T || T => T
!F || T => T 
!T || F => F 
!F || F => T


!(InBlackoutDateRange) ||
(isUserAdmin || userHasModerationPrivilages || isUserGroupTA || isCourseAdmin || isCourseStaff)

**Review by awais-ansari — CHANGES_REQUESTED** [review]

File: src/discussions/data/hooks.test.jsx [hunk 9]
```diff
@@ -82,4 +83,101 @@ describe('Hooks', () => {
...
+          },
+        });
+      });
+
+      test('return true when blackout dates are not active and Role is not Learner', async () => {
+        const { queryByText } = renderComponent();
+        expect(queryByText('true')).toBeInTheDocument();
+      });
+    });
+  });
```
Comment (mehaknasir) [inline src/discussions/data/hooks.test.jsx:182]: 1. remove beforeEach from each test function as it contains only 1 test case.
2. remove async keyword from beforeEach function
3. structural changes

` describe('blackout dates are active', () => {
      beforeEach(async () => {
        store = initializeStore({
          config: {
            blackouts: [{ start: '2022-11-25T00:00:00Z', end: '2050-11-25T23:59:00Z' }],
            hasModerationPrivileges: false,
            isGroupTa: false,
            isCourseAdmin: false,
            isCourseStaff: false,
            isUserAdmin: false,
          },
        });
      });

      test('return false when blackout dates are active and Role is Learner', async () => {
        const { queryByText } = renderComponent();
        expect(queryByText('false')).toBeInTheDocument();
      });
       test('return false when blackout dates are active and Role is other than', async () => {
        const { queryByText } = renderComponent();
        expect(queryByText('false')).toBeInTheDocument();
      });
    });`


` describe('blackout dates are inActive', () => {
      beforeEach(async () => {
        store = initializeStore({
          config: {
            blackouts: [{ start: '2022-11-25T00:00:00Z', end: '2050-11-25T23:59:00Z' }],
            hasModerationPrivileges: false,
            isGroupTa: false,
            isCourseAdmin: false,
            isCourseStaff: false,
            isUserAdmin: false,
          },
        });
      });

      test('return {{true}} when blackout dates are active and Role is Learner', async () => {
       
      });
       test('return {{true}} when blackout dates are active and Role is not Learner', async () => {
       
      });
    });`

**Review by mehaknasir — CHANGES_REQUESTED** [review]

**Review by awais-ansari — COMMENTED** [review]
I agree with @mehaknasir concerns. Other than that PR LGTM.

**Review by awais-ansari — APPROVED** [review]

**Pull request merged**